blaze/expr/math.py

```diff
@@ -11,67 +11,166 @@
 
 __all__ = ['abs', 'sqrt', 'sin', 'sinh', 'cos', 'cosh', 'tan', 'tanh', 'exp',
            'expm1', 'log', 'log10', 'log1p', 'acos', 'acosh', 'asin', 'asinh',
-           'atan', 'atanh', 'radians', 'degrees', 'ceil', 'floor', 'trunc',
-           'isnan', 'notnull', 'RealMath', 'IntegerMath', 'BooleanMath',
-           'Math']
+           'atan', 'atanh', 'radians', 'degrees', 'atan2', 'ceil', 'floor',
+           'trunc', 'isnan', 'notnull', 'UnaryMath', 'BinaryMath',
+           'greatest', 'least']
 
 
-class Math(UnaryOp):
-    pass
-
+class UnaryMath(UnaryOp):
 
-class RealMath(Math):
     """Mathematical unary operator with real valued dshape like sin, or exp
     """
+    __slots__ = '_hash', '_child'
+
     _dtype = real
 
 
-class abs(RealMath): pass
-class sqrt(RealMath): pass
+class BinaryMath(BinOp):
+    __slots__ = '_hash', 'lhs', 'rhs'
+    __inputs__ = 'lhs', 'rhs'
 
-class sin(RealMath): pass
-class sinh(RealMath): pass
-class cos(RealMath): pass
-class cosh(RealMath): pass
-class tan(RealMath): pass
-class tanh(RealMath): pass
+    _dtype = real
 
-class exp(RealMath): pass
-class expm1(RealMath): pass
-class log(RealMath): pass
-class log10(RealMath): pass
-class log1p(RealMath): pass
+    def __str__(self):
+        return '%s(%s, %s)' % (type(self).__name__, self.lhs, self.rhs)
 
-class acos(RealMath): pass
-class acosh(RealMath): pass
-class asin(RealMath): pass
-class asinh(RealMath): pass
-class atan(RealMath): pass
-class atanh(RealMath): pass
 
-class radians(RealMath): pass
-class degrees(RealMath): pass
+class abs(UnaryMath):
```

Comment:
Could we programmtically generate these?

```python
for name in unary_type_names:
    locals()[name] = type(name, (UnaryMath,), {})
```
same for the binary names.
This makes it easier to scan through quickly imho.